File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/VerifyCmd.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.Verification;
import sop.cli.picocli.Print;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.DetachedVerify;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

@CommandLine.Command(name = "verify",
        resourceBundle = "msg_detached-verify",
        exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
public class VerifyCmd extends AbstractSopCmd {

    @CommandLine.Parameters(index = "0",
            paramLabel = "SIGNATURE")
    String signature;

    @CommandLine.Parameters(index = "1..*",
            arity = "1..*",
            paramLabel = "CERT")
    List<String> certificates = new ArrayList<>();

    @CommandLine.Option(names = {"--not-before"},
            paramLabel = "DATE")
    String notBefore = "-";

    @CommandLine.Option(names = {"--not-after"},
            paramLabel = "DATE")
    String notAfter = "now";

    @Override
    public void run() {
        DetachedVerify detachedVerify = throwIfUnsupportedSubcommand(
                SopCLI.getSop().detachedVerify(), "verify");

        if (notAfter != null) {
            try {
                detachedVerify.notAfter(parseNotAfter(notAfter));
            } catch (SOPGPException.UnsupportedOption unsupportedOption) {
                String errorMsg = getMsg("sop.error.feature_support.option_not_supported", "--not-after");
                throw new SOPGPException.UnsupportedOption(errorMsg, unsupportedOption);
            }
        }
        if (notBefore != null) {
            try {
                detachedVerify.notBefore(parseNotBefore(notBefore));
            } catch (SOPGPException.UnsupportedOption unsupportedOption) {
                String errorMsg = getMsg("sop.error.feature_support.option_not_supported", "--not-before");
                throw new SOPGPException.UnsupportedOption(errorMsg, unsupportedOption);
            }
        }

        for (String certInput : certificates) {
            try (InputStream certIn = getInput(certInput)) {
                detachedVerify.cert(certIn);
            } catch (IOException ioException) {
                throw new RuntimeException(ioException);
            } catch (SOPGPException.BadData badData) {
                String errorMsg = getMsg("sop.error.input.not_a_certificate", certInput);
                throw new SOPGPException.BadData(errorMsg, badData);
            }
        }

        if (signature != null) {
            try (InputStream sigIn = getInput(signature)) {
                detachedVerify.signatures(sigIn);
            } catch (IOException e) {
                throw new RuntimeException(e);
            } catch (SOPGPException.BadData badData) {
                String errorMsg = getMsg("sop.error.input.not_a_signature", signature);
                throw new SOPGPException.BadData(errorMsg, badData);
            }
        }

        List<Verification> verifications;
        try {
            verifications = detachedVerify.data(System.in);
        } catch (SOPGPException.NoSignature e) {
            String errorMsg = getMsg("sop.error.runtime.no_verifiable_signature_found");
            throw new SOPGPException.NoSignature(errorMsg, e);
        } catch (IOException ioException) {
            throw new RuntimeException(ioException);
        } catch (SOPGPException.BadData badData) {
            String errorMsg = getMsg("sop.error.input.stdin_not_a_message");
            throw new SOPGPException.BadData(errorMsg, badData);
        }

        for (Verification verification : verifications) {
            Print.outln(verification.toString());
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/VerifyCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import picocli.CommandLine.*
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException.*

@Command(
    name = "verify",
    resourceBundle = "msg_detached-verify",
    exitCodeOnInvalidInput = UnsupportedOption.EXIT_CODE)
class VerifyCmd : AbstractSopCmd() {

    @Parameters(index = "0", paramLabel = "SIGNATURE") lateinit var signature: String

    @Parameters(index = "1..*", arity = "1..*", paramLabel = "CERT")
    lateinit var certificates: List<String>

    @Option(names = [OPT_NOT_BEFORE], paramLabel = "DATE") var notBefore: String = "-"

    @Option(names = [OPT_NOT_AFTER], paramLabel = "DATE") var notAfter: String = "now"

    override fun run() {
        val detachedVerify =
            throwIfUnsupportedSubcommand(SopCLI.getSop().detachedVerify(), "verify")

        throwIfUnsupportedOption(OPT_NOT_AFTER) { detachedVerify.notAfter(parseNotAfter(notAfter)) }
        throwIfUnsupportedOption(OPT_NOT_BEFORE) {
            detachedVerify.notBefore(parseNotBefore(notBefore))
        }

        for (certInput in certificates) {
            try {
                getInput(certInput).use { certIn -> detachedVerify.cert(certIn) }
            } catch (ioException: IOException) {
                throw RuntimeException(ioException)
            } catch (badData: BadData) {
                val errorMsg = getMsg("sop.error.input.not_a_certificate", certInput)
                throw BadData(errorMsg, badData)
            }
        }

        try {
            getInput(signature).use { sigIn -> detachedVerify.signatures(sigIn) }
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (badData: BadData) {
            val errorMsg = getMsg("sop.error.input.not_a_signature", signature)
            throw BadData(errorMsg, badData)
        }

        val verifications =
            try {
                detachedVerify.data(System.`in`)
            } catch (e: NoSignature) {
                val errorMsg = getMsg("sop.error.runtime.no_verifiable_signature_found")
                throw NoSignature(errorMsg, e)
            } catch (ioException: IOException) {
                throw RuntimeException(ioException)
            } catch (badData: BadData) {
                val errorMsg = getMsg("sop.error.input.stdin_not_a_message")
                throw BadData(errorMsg, badData)
            }

        for (verification in verifications) {
            println(verification.toString())
        }
    }

    companion object {
        const val OPT_NOT_BEFORE = "--not-before"
        const val OPT_NOT_AFTER = "--not-after"
    }
}
